src/SpirvIR.cpp

```diff
@@ -0,0 +1,1710 @@
+}
+
+SpvId SpvInstruction::result_id() const {
+    user_assert(is_defined()) << "An SpvInstruction must be defined before accessing its properties\n";
+    return contents->result_id;
+}
+
+SpvId SpvInstruction::type_id() const {
+    user_assert(is_defined()) << "An SpvInstruction must be defined before accessing its properties\n";
+    return contents->type_id;
+}
+
+SpvOp SpvInstruction::op_code() const {
+    user_assert(is_defined()) << "An SpvInstruction must be defined before accessing its properties\n";
+    return contents->op_code;
+}
+
+SpvId SpvInstruction::operand(uint32_t index) {
+    user_assert(is_defined()) << "An SpvInstruction must be defined before accessing its properties\n";
+    return contents->operands[index];
+}
+
+bool SpvInstruction::has_type() const {
+    if (!is_defined()) { return false; }
+    return contents->type_id != SpvNoType;
+}
+
+bool SpvInstruction::has_result() const {
+    if (!is_defined()) { return false; }
+    return contents->result_id != SpvNoResult;
+}
+
+bool SpvInstruction::is_defined() const {
+    return contents.defined();
+}
+
+bool SpvInstruction::is_immediate(uint32_t index) const {
+    user_assert(is_defined()) << "An SpvInstruction must be defined before modifying its properties\n";
+    return contents->immediates[index];
+}
+
+uint32_t SpvInstruction::length() const {
+    user_assert(is_defined()) << "An SpvInstruction must be defined before modifying its properties\n";
+    return (uint32_t)contents->operands.size();
+}
+
+SpvBlock SpvInstruction::block() const {
+    user_assert(is_defined()) << "An SpvInstruction must be defined before modifying its properties\n";
+    return contents->block;
+}
+
+void SpvInstruction::add_data(uint32_t bytes, const void *data) {
+    uint32_t extra_words = (bytes + 3) / 4;
+    const uint8_t *ptr = (const uint8_t *)data;
+    size_t bytes_copied = 0;
+    for (uint32_t i = 0; i < extra_words; i++) {
+        size_t copy_size = std::min(bytes - bytes_copied, (size_t)4);
+        SpvId entry = 0;
+        memcpy(&entry, ptr, copy_size);
+        bytes_copied += copy_size;
+        add_immediate(entry);
+        ptr++;
+    }
+}
+
+void SpvInstruction::add_string(const std::string &str) {
+    add_data(str.length() + 1, (const void *)str.c_str());
+}
+
+void SpvInstruction::encode(SpvBinary &binary) const {
+    user_assert(is_defined()) << "An SpvInstruction must be defined before accessing its properties\n";
+
+    // Count the number of 32-bit words to represent the instruction
+    uint32_t word_count = 1;
+    word_count += has_type() ? 1 : 0;
+    word_count += has_result() ? 1 : 0;
+    word_count += length();
+
+    // Preface the instruction with the format
+    // - high 16-bits indicate instruction length (number of 32-bit words)
+    // - low 16-bits indicate op code
+    binary.push_back(((word_count) << SpvWordCountShift) | contents->op_code);
+    if (has_type()) { binary.push_back(contents->type_id); }
```

Comment:
Style nit, here and elsewhere: if statements should never be on the same line like this (I'm suprised that clang-format didn't flag this)
```
    if (has_type()) { 
        binary.push_back(contents->type_id); 
    }
```